### DISCUSSION TITLE
Leveraging multiple runners on GitHub actions

### DISCUSSION THREAD

**Discussion by sppatel** [vercel/turborepo#1056] [Help] [answered]
Turbo repo as a task runner appears to be most efficient when running under a single CI runner having visibility and control over all tasks.  However CI environments such as GitHub actions allow you to paralyze jobs across multiple machines.  In this case, your workflow jobs would declare dependencies on each other (if one job depended on one or more other jobs).  This conflicts with the dependencies declared in turbo.json.

So what is the recommendation for CI? To run all tasks under a single machine?  Or is there a way to efficiently leverage BOTH turbo repo's parallelization along side a CI's parallelization support across different machines.  I'm not sure how the latter would be possible because each parallel CI job would essentially be executing a new turbo process.

Perhaps if you choose to leverage CI parallelization you are forgoing turbo's or limiting parallelization but maybe the cache can be shared for the entire workflow?  Is that the correct thought process?

Any guidance would be appreciated.

**Comment by anthonyshew** [comment, marked as answer]
Answering this question is likely specific to your own use case and what you'd like to implement.

`turbo`'s design keeps it simple: tell us what to run and we will parallelize it. We understand your project's dependency graph and we'll do the work required to make it happen as quickly as possible through whatever combination of caching and parallelization is needed.

GitHub Actions also have these capabilities but they're certainly not as simple as `turbo build`. I'd encourage you to explore `turbo`'s abilities first and then see if splitting up jobs further is needed.
